pkg/controller/cert.go

```diff
@@ -0,0 +1,208 @@
+	namespace, secretName string,
+	caKeyPair *triple.KeyPair,
+	commonName string,
+	organizations []string,
+	owner *metav1.OwnerReference,
+	serverCACert *x509.Certificate) (*triple.KeyPair, error) {
+	keyPair, err := GetKeyPair(client, namespace, secretName)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error getting server key pair")
+	}
+
+	if keyPair != nil {
+		glog.Infof("Retrieved client key/cert %s from kubernetes", organizations)
+		return keyPair, nil
+	}
+
+	keyPair, err = triple.NewClientKeyPair(caKeyPair, commonName, organizations)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating client key pair")
+	}
+
+	if err = SaveKeyPair(client, namespace, secretName, keyPair, owner, serverCACert); err != nil {
+		return nil, errors.Wrap(err, "Error saving client key pair")
+	}
+
+	return keyPair, nil
+}
+
+// GetKeyPair will return the KeyPair if it exists
+func GetKeyPair(client kubernetes.Interface, namespace, secretName string) (*triple.KeyPair, error) {
+	var keyPair triple.KeyPair
+
+	secret, err := client.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrap(err, "Error getting secret")
+	}
+
+	if bytes, ok := secret.Data[PrivateKeyKeyName]; ok {
+		obj, err := cert.ParsePrivateKeyPEM(bytes)
+		if err != nil {
+			return nil, errors.Wrap(err, "Error parsing secret")
+		}
+
+		key, ok := obj.(*rsa.PrivateKey)
+		if !ok {
+			return nil, errors.New("Invalid pem format")
+		}
+		keyPair.Key = key
+	}
+
+	if bytes, ok := secret.Data[CertKeyName]; ok {
+		certs, err := cert.ParseCertsPEM(bytes)
+		if err != nil || len(certs) != 1 {
+			return nil, errors.Errorf("Cert parse error %s, %d", err, len(certs))
+		}
+		keyPair.Cert = certs[0]
+	}
+
+	return &keyPair, nil
+}
+
+// SaveKeyPair saves a private key and cert to kubernetes
+func SaveKeyPair(client kubernetes.Interface, namespace, secretName string, keyPair *triple.KeyPair, owner *metav1.OwnerReference, caCert *x509.Certificate) error {
+	privateKeyBytes := cert.EncodePrivateKeyPEM(keyPair.Key)
+	certBytes := cert.EncodeCertPEM(keyPair.Cert)
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: namespace,
+		},
+		Type: "Opaque",
+		Data: map[string][]byte{
+			PrivateKeyKeyName: privateKeyBytes,
+			CertKeyName:       certBytes,
+		},
+	}
+
+	if owner != nil {
+		secret.OwnerReferences = []metav1.OwnerReference{*owner}
+	}
+
+	if caCert != nil {
+		caCertBytes := cert.EncodeCertPEM(caCert)
+		secret.Data[CACertKeyName] = caCertBytes
+	}
+
+	_, err := client.CoreV1().Secrets(namespace).Create(secret)
+	if err != nil {
+		return errors.Wrap(err, "Error creating cert")
+	}
+
+	return nil
+}
+
+// GetNamespace returns the namespace the pod is executing in
+func GetNamespace() string {
+	if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
```

Comment:
how does this hard-coded path work in windows clusters?

Reply:
No idea!  But how would it matter?  We are building linux containers?